WikiWatershed/model-my-watershed
# Add user profile form

## Overview

Adds a second form shown after logging in for the first time that prompts for additional, optional information.

Connects #2255 

### Demo

<img width="844" alt="screen shot 2017-10-25 at 2 38 16 pm" src="https://user-images.githubusercontent.com/17363/32024714-639cf7fa-b992-11e7-9c00-02a7ca206a64.png">

### Notes

Following the pattern of `ItsiUser` I have created a new model related to `User` to hold additional, optional information.

- All additional user profile fields are optional.
- `user_type` is selected from a defined list of choices and defaults to "Unspecified."
- `country` is selected from a defined list of choices and defaults to "US."

Based on the style of the login modal form, I created a new user profile modal form

- The form is shown after logging in only if both the `was_skipped` and `is_complete` flags on the `UserProfile` are false.
- Making the decision whether or not to show the form required passing the login response as an argument through the chain of success event handlers.
- The options for the `country` and `user_type` fields are passed from Python to Javascript via the existing `window.clientSettings` object.
- The nunjucks template is built at compile time, not at run time, so there is some Javascript that completes the construction of the option tags on the client.
- I edited `bundle.sh` to include the `bootstrap-select` stylesheet that was previously added to the project.

## Testing Instructions

 * Run migrations and rebuild the vendor bundle.
 * Log in and verify that the new form is displayed.
 * Click "I'll do this later." Log out, log back in, and verify that the new form is not shown.
 * Log in as a different user.
 * Fill out the form and click "Submit profile"
 * Log out, log back in, and verify that the new form is not shown.
 * Run './scripts/manage.sh dbshell` and verify the profile data
```sql
select * from user_userprofile;
```
